Add a test to check bad use of application owned memory

marius-pelegrin-arm opened the pull request:
Add the test vulkan_pnext_chain that checks the use of application owned memory passed through pNext chains. A bug has been found in GFXReconstruct where it uses application owned pointer after the vulkan call that gave the pointer, thus with application having the possibility of freeing the underlying memory. This unsafe use causes a SEGFAULT. This is a minimal repro of this bug.

*Warning*: To be usable, this test needs a VkDebugUtilsLabelEXT to be recognized as an end of frame by GFXReconstruct. This is not the case at the moment.

marius-pelegrin-arm committed:
Add a test to check bad use of application owned memory

Add the test vulkan_pnext_chain that checks the use of application
owned memory passed through pNext chains. A bug has been found in
GFXReconstruct where it uses application owned pointer after the
vulkan call that gave the pointer, thus with application having the
possibility of freeing the underlying memory. This unsafe use causes
a SEGFAULT. This is a minimal repro of this bug.

per-mathisen-arm commented on src/vulkan_pnext_chain.cpp:
```diff
@@ -0,0 +1,188 @@
+	debugLabel.color[2] = 0.f;
+	debugLabel.color[3] = 1.f;
+
+	PFN_vkCmdInsertDebugUtilsLabelEXT pfnVkCmdInsertDebugUtilsLabelEXT = (PFN_vkCmdInsertDebugUtilsLabelEXT) vkGetInstanceProcAddr(vulkan.instance, "vkCmdInsertDebugUtilsLabelEXT");
```
Can you add it to the vulkan object, see src/vulkan_common.cpp line 441 ?

marius-pelegrin-arm replied:
Done.

per-mathisen-arm commented on src/vulkan_pnext_chain.cpp:
```diff
@@ -0,0 +1,188 @@
+
+	/* Where the magic happens...
+
+	Those lines check that the trace tool does not save and reuse application owned memory
```
I would remove the second sentence since it would hopefully become false soon.

marius-pelegrin-arm replied:
Removed, and changed a bit the comment to make it more explicit.

per-mathisen-arm commented on src/vulkan_pnext_chain.cpp:
```diff
@@ -0,0 +1,188 @@
+	deviceGroupInfo.sType = VK_STRUCTURE_TYPE_BIND_IMAGE_MEMORY_DEVICE_GROUP_INFO;
+	deviceGroupInfo.pNext = nullptr;
+	deviceGroupInfo.deviceIndexCount = 1;
+	deviceGroupInfo.pDeviceIndices = nullptr;
```
Why not just point this to an array of size 1 with a zero in it here, rather than allocate and delete below?

marius-pelegrin-arm replied:
That's the whole point of the test.

per-mathisen-arm replied:
Ohhh, right. I get it now :-)

marius-pelegrin-arm committed:
Requested changes

- Add vkCmdInsertDebugUtilsLabel to vulkan_setup_t
- Hopefully better comment to explain what the test does
